### PR TITLE
clean up nss for state based slicing

### PR DESCRIPTION
**Pull request by yallup**
Additional code needed for standalone slice after the current API changes, makes the slice direction proposal take current position as argument, purely to generate shapes, in line with blackjax.util.generate_xyz

ns tests still broken but from a high level now just lift the generate slice direction directly so should be written to test this dict state explicitly

### Summary
  Simplified the nested slice sampling (NSS) implementation by moving from PyTree-based covariance storage to simpler
  matrix-based approach, aligning the direction generation API with the slice sampler's state-based interface.

  ### Key Changes

  **Slice Sampler (`blackjax/mcmc/ss.py`)**
  - Updated `build_hrss_kernel` signature: `generate_slice_direction_fn` now accepts `(rng_key, position)` instead of
  just `(rng_key)`
  - Refactored `sample_direction_from_covariance` to:
    - Accept `position` parameter to extract shape information
    - Sample from standard normal distribution and normalize using covariance inverse
    - Return direction in original position's PyTree structure

  **Nested Slice Sampling (`blackjax/ns/nss.py`)**
  - Removed redundant `sample_direction_from_covariance` wrapper function (67 lines deleted)
  - Simplified `compute_covariance_from_particles`: now returns covariance matrix directly instead of PyTree structure
  - Updated kernel to pass `state.position` and `params["cov"]` to direction generation function
  - Removed unused imports (`ravel_pytree`, `get_first_row`, `particles_as_rows`)

  **Tests (`tests/mcmc/test_slice_sampling.py`)**
  - Updated direction function signatures to accept `position` parameter

  ### Impact
  - Cleaner API with explicit state dependency
  - Reduced code complexity by ~50 lines
  - Maintains same functionality with simpler data flow

### CODE REVIEW

File: blackjax/mcmc/ss.py [hunk 1]
```diff
@@ -317,11 +322,12 @@ def sample_direction_from_covariance(rng_key: PRNGKey, cov: Array) -> Array:
     Array
         A normalized direction vector (1D array).
     """
-    d = jax.random.multivariate_normal(rng_key, mean=jnp.zeros(cov.shape[0]), cov=cov)
+    p, unravel_fn = ravel_pytree(position)
+    d = jax.random.normal(rng_key, shape=p.shape, dtype=p.dtype)
```
Comment (williamjameshandley) [inline blackjax/mcmc/ss.py]: either missing a multiply by cholesky, or should use the multivariate_normal version.

File: blackjax/ns/nss.py [hunk 2]
```diff
@@ -175,10 +123,7 @@ def compute_covariance_from_particles(
         This means each leaf of `cov_pytree` will have a shape `(D, *leaf_original_dims)`.
     """
     cov_matrix = jnp.atleast_2d(particles_covariance_matrix(state.particles))
-    single_particle = get_first_row(state.particles)
-    _, unravel_fn = ravel_pytree(single_particle)
-    cov_pytree = jax.vmap(unravel_fn)(cov_matrix)
-    return {"cov": cov_pytree}
+    return {"cov": cov_matrix}
```
Comment (williamjameshandley) [inline blackjax/ns/nss.py]: could possibly make it a one-liner

Comment (williamjameshandley) [reply blackjax/ns/nss.py]: and also possibly reduce docstring length for consistency with the rest of blackjax?

File: blackjax/ns/nss.py [hunk 3]
```diff
@@ -242,7 +187,7 @@ def inner_kernel(
         rng_key, state, logprior_fn, loglikelihood_fn, loglikelihood_0, params
     ):
         rng_key, prop_key = jax.random.split(rng_key, 2)
-        d = generate_slice_direction_fn(prop_key, params)
+        d = generate_slice_direction_fn(prop_key, state.position, **params)
```
Comment (williamjameshandley) [inline blackjax/ns/nss.py:117]: unclear whether we need the ** I've put in here...

**Review by williamjameshandley — CHANGES_REQUESTED** [review]
This is much nicer post refactor.

File: blackjax/mcmc/ss.py (hunk 1 above)
Comment (yallup) [reply blackjax/mcmc/ss.py]: Put in a more stable version that uses the cholesky anyway which we’ve discussed. I think it caches it so this is the same as “precomputing” the cholesky, although perhaps should make sure this is true!

One outstanding Question was that we had at one point:

```python 
cov_matrix = jnp.atleast_2d(particles_covariance_matrix(state.particles))
cov_matrix *= cov_matrix.shape[0] + 2
```

As an optimal cov, can study what is optimal empirically but good to fold this potential change in here at this point if the above is in theory optimal

File: blackjax/ns/nss.py (hunk 2 above)
Comment (yallup) [reply blackjax/ns/nss.py]: In general in favour of this, especially for simple functions. Only things where I think we want verbose docstrings are for top level api documentation, or functions that are not necessarily  clear such as delete_fn

**Comment by williamjameshandley** [comment]
## Updates addressing review comments

I've made several improvements based on the review feedback:

### 1. Docstring improvements

- **Simplified `compute_covariance_from_particles`**: Reduced to a concise one-liner as suggested
- **Clarified `generate_slice_direction_fn` signature**: Changed from `**params` to `**kwargs` and explained that keyword arguments are unpacked from the dict returned by `adapt_direction_params_fn`

### 2. Optimal covariance scaling: 2√(d+2) factor

Added the theoretically optimal scaling factor for slice sampling direction proposals. This addresses the outstanding question about the `cov_matrix *= cov_matrix.shape[0] + 2` scaling.

#### Mathematical derivation

For points uniformly distributed in a d-ball of radius R, the empirical covariance is:

**Σ_empirical = (R²/(d+2)) I**

This follows from computing E[x_i²]:
- By spherical symmetry: Σ = σ² I where σ² = E[r²]/d
- Radial density: p(r) = d r^(d-1) / R^d for 0 ≤ r ≤ R
- E[r²] = ∫₀^R r² · (d r^(d-1)/R^d) dr = dR²/(d+2)
- Therefore: σ² = R²/(d+2)

#### The scaling factor breakdown

The optimal **2√(d+2)** scaling on the Cholesky factor (equivalent to **4(d+2)** on covariance) corrects two issues:

1. **√(d+2) factor - bias correction**: The empirical covariance underestimates spatial extent by (d+2). To recover the true scale R from variance R²/(d+2), we scale by √(d+2).

2. **Factor of 2 - diameter vs radius**: Slice sampling's initial stepping-out interval should span the full diameter (2R), not just radius (R). This requires scaling directions by 2.

Combined: **L_optimal = 2√(d+2) · L_empirical**

#### Implementation

- Applied scaling in `sample_direction_from_covariance` after Cholesky decomposition
- Kept empirical covariance computation pure (no scaling in `compute_covariance_from_particles`)
- Added detailed docstring explaining the mathematics

#### Validation

Created `validate_covariance.py` demonstrating the Σ = R²/(d+2) I formula with <0.03% error across dimensions 2-20.

### 3. Confirmed `**params` unpacking

Keeping the `**params` unpacking as-is for extensibility. This allows users to create custom direction functions with additional parameters beyond just `cov`.

---

All changes committed in 667fc63fb6ca8c9c032e6005dad5b96ded1b94a5

**Review by williamjameshandley — APPROVED** [review]
This is much nicer post refactor.
